### ISSUE TITLE
Combining UseBindPortIP and loadbalancer service port

### ISSUE THREAD

**Issue by sirlatrom** [containous/traefik#6187]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

This is a feature request for the Docker provider (without Swarm Mode).

When enabling `UseBindPortIP`, as per the [documentation](https://docs.traefik.io/providers/docker/#port-detection), the first port in the Docker API response is chosen. Then there's the [`traefik.http.services.myservice.loadbalancer.server.port`](https://docs.traefik.io/routing/providers/docker/#port) option which allows you to inform Traefik what port to use for a server. _Without_ `UseBindPortIP`, it makes sense that the resulting Traefik server will simply use the port given in the `server.port` option. The documentation says,

> Registers a port. Useful when the container exposes multiples ports. Mandatory for Docker Swarm (see the section ["Port Detection with Docker Swarm"](https://docs.traefik.io/providers/docker/#port-detection_1)).

, however it is not clear from that how Traefik is expected to behave when `UseBindPortIP` is enabled. I would have thought the host port to be used, and the `server.port` option just helping Traefik to choose the right one based on the container port that matches the `server.port` option.

I realise such a change would maybe be considered breaking as people may have relied on the current behaviour in Traefik 2.0+, but on the other hand there is currently no way to help Traefik choose the right port when there are multiple.

I'm not sure if this is really a bug like #3622 was for pre 2.0.

**Comment by ldez** [comment]
Hello,

The behavior of `useBindPortIP` is explained in the section "examples".

https://docs.traefik.iov/2.1/providers/docker/#usebindportip

**Comment by sirlatrom** [comment]
Right, so I _think_ I'm referring to the highlighted row in the docs, but maybe that's my misunderstanding:

| port label | Container's binding | Routes to |
| -- | -- | -- |
| - | - | IntIP:IntPort |
| - | ExtPort:IntPort | IntIP:IntPort |
| - | ExtIp:ExtPort:IntPort | ExtIp:ExtPort |
| LblPort | - | IntIp:LblPort |
| **LblPort** | **ExtIp:ExtPort:LblPort** | **ExtIp:ExtPort** |
| LblPort | ExtIp:ExtPort:OtherPort | IntIp:LblPort |
| LblPort | ExtIp1:ExtPort1:IntPort1 & ExtIp2:LblPort:IntPort2 | ExtIp2:LblPort |

Given the following `NetworkSettings.Ports` block, and a `server.port` value of `8080`, should the result not then be `10.107.171.68:32832`? Or am I misinterpreting the table?

```json
            "Ports": {
                "80/tcp": null,
                "8080/tcp": [
                    {
                        "HostIp": "10.107.171.68",
                        "HostPort": "32832"
                    }
                ]
            },
```

**Comment by sirlatrom** [comment]
I believe this code governs the use of `server.port` port label:

https://github.com/containous/traefik/blob/72e702a15a7a68481e7f2d65714046fa1124d93a/pkg/provider/docker/config.go#L164-L167

It does not take `UseBindPortIP` into account, so how the external port could ever be used with a port label being set is beyond me.

**Comment by sirlatrom** [comment]
There are two scenarios:

1. You're lucky and your desired port number is the lowest, and things just work.
2. Otherwise, then you get one of:
    a. The lower port number is used instead of the ExtPort which is mapped to IntPort, or
    b. You specify LblPort and you get that, verbatim, instead of ExtPort

Here's a script to play with the behaviour. Change the `--expose 90` to `--expose 70` to try scenario 2b, remove the `server.port` label to try scenario 2a.

Please note that this is with an explicitly published IP (set in the `EXT_IP` variable) for the port, i.e. _not_ `0.0.0.0`.

```bash
#!/bin/bash

set -euo pipefail

: "${EXT_IP?Please explicitly set your external IP in EXT_IP}"

# Cleanup
trap 'set +e; docker container kill triage-6187; kill $pid || killall traefik' EXIT

# Run whoami container detached
docker container run \
    --label traefik.http.routers.triage-6187.rule='Path(`/triage-6187`)' \
    --label traefik.http.services.triage-6187.loadbalancer.server.port=80 \
    --name triage-6187 \
    --rm \
    --detach \
    --expose 90 \
    --publish \
    "${EXT_IP}::80" \
    containous/whoami:dev &&
    docker container inspect --format '{{json .NetworkSettings.Ports}}' triage-6187 | jq .

# Run Traefik in the background
env -i "$(command -v traefik)" \
    --entrypoints.http.address=:8000 \
    --api.insecure=true \
    --api.dashboard \
    --providers.docker.usebindportip &
pid=$!

# Wait for Traefik to discover container
sleep 2

# Get service from API
curl -f -s -S http://localhost:8080/api/http/services/triage-6187@docker | jq '.loadBalancer.servers|first.url'

# Test acces to container through Traefik
curl -f -s S http://localhost:8000/triage-6187 || echo

# Give some time for manual inspection
sleep 20
```

**Comment by juliens** [comment]
Hi @sirlatrom ,

I tried your "highlight case in the table" and I can confirm that it doesn't work. A fix is coming soon.

**Comment by traefiker** [comment]
Closed by #6199.
